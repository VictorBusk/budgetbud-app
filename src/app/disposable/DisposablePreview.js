import React, {Component} from 'react';
import {FlatList, View} from 'react-native';
import {
    Container,
    Content,
    Button,
    Spinner,
    ListItem,
    Body,
    Right,
    Text
} from "native-base";
import _ from 'lodash';
import {AppHeader, Separator} from "../../components";
import {connect} from "react-redux";
import I18n from "../../strings/i18n";
import {button, text, container, color} from "../../style";
import {showWarningToast} from "../../helpers";
import {editDisposable, resetDisposableError, getCategories} from "../../redux/actions";

class DisposablePreview extends Component {
    componentWillReceiveProps(nextProps) {
        if (nextProps.disposableError) {
            showWarningToast(nextProps.disposableError);
            this.props.resetDisposableError();
        }
    }

    onSavePress = async () => {
        if (this.props.disposableLoading)
            return;

        await this.props.editDisposable(this.props);
        this.props.navigation.pop(2);
        this.props.getCategories(this.props.budgetID);
    };

    render() {
        return (
            <Container style={container.signedInContainer}>
                <AppHeader headerText={I18n.t('disposablePreviewHeader')}
                           showBackButton={true}
                           onLeftButtonPress={() => {
                               if (!this.props.disposableLoading) {
                                   this.props.navigation.pop();
                               }
                           }}/>

                <Content style={{flex: 4}}>
                    <FlatList
                        data={this.props.categoryDisposableItems}
                        renderItem={this.renderItem}
                        keyExtractor={item => item.categoryID}
                    />
                </Content>

                <Separator/>

                <Button rounded
                        onPress={() => this.onSavePress()}
                        style={[button.defaultButton, color.button]}
                >
                    {this.props.disposableLoading ? (
                        <Spinner color='#D0D0D0'/>) : (
                        <Text style={text.submitButtonText}>
                            {I18n.t('disposablePreviewSaveButton')}
                        </Text>
                    )}
                </Button>
            </Container>
        );
    }

    renderItem = ({item}) => {
        return (
            <ListItem>
                <View style={container.fullWidth}>
                    <Body>
                    <Text style={color.text}>{item.name}</Text>
                    </Body>
                    <View style={[container.removeIndenting, {flexDirection: 'row'}]}>
                        <Body style={{flex: 1}}>
                        <Text note>{I18n.t('disposablePreviewBefore')}</Text>
                        <Text note>{I18n.t('disposablePreviewAfter')}</Text>
                        </Body>
                        <Right style={{flex: 2}}>
                            <Text note>{item.beforeAmount} {I18n.t('currency')}</Text>
                            <Text note>{item.afterAmount} {I18n.t('currency')} (
                                {(item.beforeAmount < item.afterAmount) ? '+' : ''}
                                {item.amountDiff} {I18n.t('currency')})
                            </Text>
                        </Right>
                    </View>
                </View>
            </ListItem>
        );
    }
}

const mapStateToProps = (state) => {
    const budgetID = state.budget.budgetID;
    const {
        categories,
        selectedCategories,
    } = state.category;
    const {
        disposableCategorySubtractions,
        disposableLoading,
        disposable,
        tmpDisposable
    } = state.disposable;

    const categoryDisposableItems = _.map(selectedCategories, (item, key) => {
        const categorySubtraction = disposableCategorySubtractions.filter((obj) => {
            return obj.categoryID === item.toString();
        });

        const category = categories.filter((obj) => {
            return obj.categoryID === item.toString();
        });

        const amountToSubtract = categorySubtraction[0].amountDifference;
        let amountDiff = amountToSubtract * (-1);
        let categoryOfDebtAmount = 0;

        const beforeAmount = category[0].amount + categoryOfDebtAmount;

        return {
            amountDiff: amountDiff,
            beforeAmount: beforeAmount,
            afterAmount: (beforeAmount - amountToSubtract),
            name: category[0].name,
            categoryID: item
        };
    });

    categoryDisposableItems.push({
        amountDiff: (tmpDisposable - disposable),
        beforeAmount: disposable,
        afterAmount: tmpDisposable,
        name: I18n.t('disposable'),
        categoryID: 'disposable'
    });

    return {
        tmpDisposable,
        categoryDisposableItems,
        disposableLoading,
        budgetID
    };
};

const mapDispatchToProps = {
    editDisposable,
    resetDisposableError,
    getCategories
};

export default connect(mapStateToProps, mapDispatchToProps)(DisposablePreview);
