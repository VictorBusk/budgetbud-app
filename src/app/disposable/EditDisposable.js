import React, {Component} from 'react';
import {Container} from 'native-base';
import {connect} from "react-redux";
import I18n from "../../strings/i18n";
import {AppHeader, DisposableForm} from "../../components";
import {container} from "../../style";
import {showWarningToast} from "../../helpers";
import {checkInputAmount, correctConversion} from "../../helpers";
import {
    calculateDisposableCategoryDifferences,
    disposableChanged,
    resetDisposableError,
    resetDisposableForm,
    setTmpDisposable,
    categoriesSelected,
    getCategories
} from "../../redux/actions";

class EditDisposable extends Component {
    componentWillMount() {
        this.props.setTmpDisposable();
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.disposableError) {
            showWarningToast(nextProps.disposableError);
            this.props.resetDisposableError();
        }
    }

    onDisposableChange = (newDisposable) => {
        newDisposable = correctConversion(newDisposable);
        if (checkInputAmount(newDisposable)) {
            this.props.disposableChanged(newDisposable);
        }
    };

    onCheckBoxPress = ({categoryID}) => {
        let tmp = this.props.selectedCategories;

        if (tmp.includes(categoryID)) {
            tmp.splice(tmp.indexOf(categoryID), 1)
        } else {
            tmp.push(categoryID);
        }

        this.props.categoriesSelected(tmp);
    };

    onContinuePress = async () => {
        if (this.props.disposableCalculationLoading)
            return;

        await this.props.calculateDisposableCategoryDifferences(
            this.props.disposable,
            this.props.tmpDisposable,
            this.props.selectedCategories, () => {
                this.props.navigation.navigate('DisposablePreview');
            });
    };

    render() {
        return (
            <Container style={container.signedInContainer}>
                <AppHeader headerText={I18n.t('editDisposableHeader')}
                           showBackButton={true}
                           onLeftButtonPress={() => this.props.navigation.pop()}/>

                <DisposableForm disposableChanged={this.props.disposableChanged}
                                categoriesSelected={this.props.categoriesSelected}
                                categoryItems={this.props.categoryItems}
                                selectedCategories={this.props.selectedCategories}
                                budgetLoading={this.props.budgetLoading}
                                categoriesLoading={this.props.categoriesLoading}
                                disposable={this.props.disposable}
                                tmpDisposable={this.props.tmpDisposable}
                                disposableError={this.props.disposableError}
                                disposableCalculationLoading={this.props.disposableCalculationLoading}
                                disposableLoading={this.props.disposableLoading}
                                subtractionsLoading={this.props.subtractionsLoading}
                                onDisposableChanged={this.onDisposableChange}
                                onCheckBoxPressed={this.onCheckBoxPress}
                                onContinuePressed={this.onContinuePress}/>
            </Container>
        );
    }
}

const mapStateToProps = (state) => {
    const budgetLoading = state.budget.budgetLoading;
    const {
        disposable,
        tmpDisposable,
        disposableError,
        disposableLoading,
        disposableCalculationLoading
    } = state.disposable;
    const {
        categories,
        selectedCategories,
        categoriesLoading,
        subtractionsLoading,
        categoriesError
    } = state.category;

    const categoryItems = categories.filter((obj) => {
        return obj.amount > 0
    });

    return {
        budgetLoading,
        selectedCategories,
        categoryItems,
        categoriesLoading,
        subtractionsLoading,
        categoriesError,
        disposable,
        tmpDisposable,
        disposableError,
        disposableLoading,
        disposableCalculationLoading
    };
};

const mapDispatchToProps = {
    disposableChanged,
    getCategories,
    resetDisposableForm,
    calculateDisposableCategoryDifferences,
    categoriesSelected,
    setTmpDisposable,
    resetDisposableError
};

export default connect(mapStateToProps, mapDispatchToProps)(EditDisposable);
