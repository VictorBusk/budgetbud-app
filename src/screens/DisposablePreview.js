import React, {Component} from 'react';
import {FlatList, View} from 'react-native';
import {
    Container,
    Content,
    Button,
    Spinner,
    ListItem,
    Body,
    Right,
    Text
} from "native-base";
import _ from 'lodash';
import {AppHeader, Separator} from "../components/";
import {connect} from "react-redux";
import {button, text, container} from "../style/";
import {editDisposable} from "../actions";
import I18n from "../strings/i18n";

class DisposablePreview extends Component {
    onSavePress = () => {
        this.props.editDisposable(this.props, () => {
            this.props.navigation.pop(2);
        });
    };

    render() {
        return (
            <Container>
                <AppHeader headerText={I18n.t('disposablePreviewHeader')}
                           showBackButton={true}
                           onLeftButtonPress={() => this.props.navigation.pop()}/>

                <Content style={{flex: 4}}>
                    <FlatList
                        data={this.props.categoryDisposableItems}
                        renderItem={this.renderItem}
                    />
                </Content>

                <Separator/>

                <Button rounded
                        onPress={() => this.onSavePress()}
                        style={button.defaultButton}
                >
                    {this.props.loading ? (
                        <Spinner color='#D0D0D0'/>) : (
                        <Text style={text.submitButtonText}>
                            {I18n.t('disposablePreviewSaveButton')}
                        </Text>
                    )}
                </Button>
            </Container>
        );
    }

    renderItem = ({item}) => {
        return (
            <ListItem>
                <View style={container.fullWidth}>
                    <Body>
                    <Text>{item.name}</Text>
                    </Body>
                    <View style={[container.removeIndenting, {flexDirection: 'row'}]}>
                        <Body style={{flex: 1}}>
                        <Text note>{I18n.t('disposablePreviewBefore')}</Text>
                        <Text note>{I18n.t('disposablePreviewAfter')}</Text>
                        </Body>
                        <Right style={{flex: 2}}>
                            <Text note>{item.beforeAmount} {I18n.t('currency')}</Text>
                            <Text note>{item.afterAmount} {I18n.t('currency')} (-
                                {item.amountToSubtract} {I18n.t('currency')})
                            </Text>
                        </Right>
                    </View>
                </View>
            </ListItem>
        );
    }
}

const mapStateToProps = (state) => {
    const budgetID = state.budget.budgetID;
    const {disposable, loading} = state.disposable;
    const {
        categories,
        categorySubtractions,
    } = state.category;

    const categoryDisposableItems = _.map(categoriesOfDisposableIDs, (item, key) => {
        const categorySubtraction = categorySubtractions.filter((obj) => {
            return obj.categoryID === item.toString();
        });

        const category = categories.filter((obj) => {
            return obj.id === item.toString();
        });

        const amountToSubtract = categorySubtraction[0].amountToSubtract;
        const beforeAmount = category[0].categoryData.amount;

        return {
            amountToSubtract: amountToSubtract,
            beforeAmount: beforeAmount,
            afterAmount: (beforeAmount - amountToSubtract),
            name: category[0].categoryData.name,
            categoryID: item,
            key: key
        };
    });

    return {
        disposable,
        loading,
        budgetID,
    };
};

const mapDispatchToProps = {
    editDisposable
};

export default connect(mapStateToProps, mapDispatchToProps)(DisposablePreview);
