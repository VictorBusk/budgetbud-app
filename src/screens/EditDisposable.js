import React, {Component} from 'react';
import {Container} from 'native-base';
import {connect} from "react-redux";
import I18n from "../strings/i18n";
import AppHeader from "../components/AppHeader";
import DisposableForm from "../components/DisposableForm";

class EditDisposable extends Component {
    onContinuePress = () => {
        this.props.navigation.navigate('DisposablePreview');
    };

    render() {
        return (
            <Container>
                <AppHeader headerText={I18n.t('editDisposableHeader')}
                           showBackButton={true}
                           onLeftButtonPress={() => this.props.navigation.pop()}/>

                <DisposableForm categoryItems={this.props.disposable}
                                onContinuePress={this.onContinuePress}/>
            </Container>
        );
    }
}

styles = {
    textContainer: {
        width: '90%',
        alignSelf: 'center',
        alignItems: 'flex-start',
        marginVertical: 25
    }
};

const mapStateToProps = ({disposable}) => {
    return {disposable} = disposable;
};

export default connect(mapStateToProps, null)(EditDisposable);
