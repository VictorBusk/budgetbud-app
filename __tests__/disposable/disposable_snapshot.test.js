import React from 'react';
import configureStore from 'redux-mock-store';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import EditDisposable from "../../src/app/disposable/EditDisposable";
import DisposablePreview from "../../src/app/disposable/DisposablePreview";
import {INITIAL_BUDGET_STATE, INITIAL_CATEGORY_STATE, INITIAL_DISPOSABLE_STATE} from "../test_helper/initial_state";

Enzyme.configure({adapter: new Adapter()});

const middlewares = [];
const mockStore = configureStore(middlewares);

describe('Disposable', () => {
    it('renders EditDisposable as expected', () => {
        const wrapper = shallow(
            <EditDisposable/>,
            {
                context: {
                    store: mockStore({
                        disposable: INITIAL_DISPOSABLE_STATE,
                        budget: INITIAL_BUDGET_STATE,
                        category: INITIAL_CATEGORY_STATE
                    })
                }
            },
        );
        expect(wrapper.dive()).toMatchSnapshot();
    });

    it('renders DisposablePreview as expected', () => {
        const wrapper = shallow(
            <DisposablePreview/>,
            {
                context: {
                    store: mockStore({
                        disposable: INITIAL_DISPOSABLE_STATE,
                        budget: INITIAL_BUDGET_STATE,
                        category: INITIAL_CATEGORY_STATE
                    })
                }
            },
        );
        expect(wrapper.dive()).toMatchSnapshot();
    });
});
