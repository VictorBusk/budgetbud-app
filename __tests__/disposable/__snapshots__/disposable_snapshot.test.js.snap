// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Disposable renders DisposablePreview as expected 1`] = `
<Styled(Container)
  style={
    Object {
      "backgroundColor": "#ffffff",
    }
  }
>
  <AppHeader
    headerText="Debt preview"
    infoText="Here is a preview of how your changes affect your budget. Press [Save] to save these changes."
    onLeftButtonPress={[Function]}
    showBackButton={true}
  />
  <Styled(Content)
    style={
      Object {
        "flex": 4,
      }
    }
  >
    <FlatList
      data={
        Array [
          Object {
            "afterAmount": undefined,
            "amountDiff": NaN,
            "beforeAmount": 0,
            "categoryID": "disposable",
            "name": "Disposable",
          },
        ]
      }
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </Styled(Content)>
  <Separator />
  <Styled(Button)
    onPress={[Function]}
    rounded={true}
    style={
      Array [
        Object {
          "alignSelf": "center",
          "height": 40,
          "justifyContent": "center",
          "marginBottom": 20,
          "marginTop": 20,
          "width": "90%",
        },
        Object {
          "backgroundColor": "#00263A",
        },
      ]
    }
  >
    <Styled(Text)
      style={
        Object {
          "alignSelf": "center",
          "color": "white",
          "fontWeight": "600",
        }
      }
    >
      Save
    </Styled(Text)>
  </Styled(Button)>
</Styled(Container)>
`;

exports[`Disposable renders EditDisposable as expected 1`] = `
<Styled(Container)
  style={
    Object {
      "backgroundColor": "#ffffff",
    }
  }
>
  <AppHeader
    headerText="Disposable"
    infoText="Here you can edit your disposable and select which goals for expenses your changes will affect. Press [Continue] to see a preview of the changes."
    onLeftButtonPress={[Function]}
    showBackButton={true}
  />
  <DisposableForm
    budgetLoading={false}
    categoriesLoading={false}
    categoriesSelected={[Function]}
    categoryItems={Array []}
    disposable={0}
    disposableCalculationLoading={false}
    disposableChanged={[Function]}
    disposableError=""
    disposableLoading={false}
    onCheckBoxPressed={[Function]}
    onContinuePressed={[Function]}
    onDisposableChanged={[Function]}
    selectedCategories={Array []}
    subtractionsLoading={false}
  />
</Styled(Container)>
`;
